File: cos212_assignment1/tests_Queue.cpp
/*
MIT License

Copyright (c) 2017 Evert Geldenhuys

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#include <iostream>
#include <sstream>

#include "../catch/catch.hpp"
#include "utils.h"
#include "Deque.h"


SCENARIO("testing Queue::enqueue() and Queue::operator<<()") {
	/*
		Cases:
			- When null
			- When one
			- When many
			- When inserting negative
	*/

	GIVEN("an empty Queue object") {
		Queue list;

		WHEN("printing with one element") {
			list.enqueue(5);

			THEN("only the element is printed") {
				REQUIRE(getOutput(list) == "[5]");
			}
		}

		WHEN("Adding elements") {
			list.enqueue(1);
			list.enqueue(2);
			list.enqueue(3);
			list.enqueue(4);

			THEN("elements are added to the back")
			{
				REQUIRE(getOutput(list) == "[1,2,3,4]");
			}

			AND_WHEN("adding negative elements") {
				list.enqueue(-1);

				THEN("nothing happens") {
					REQUIRE(getOutput(list) == "[1,2,3,4]");
				}
			}
		}

		WHEN("printing empty list") {
			THEN("no crash") {
				REQUIRE(getOutput(list) == "[]");
			}
		}
	}
}

SCENARIO("testing Queue::dequeue()") {
	GIVEN("a Queue object with some elements") {
		Queue list;
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4); // 1 2 3 4

		WHEN("calling dequeue") {
			int result = list.dequeue();

			THEN("the first element is removed and returned") {
				REQUIRE(result == 1);
				REQUIRE("[2,3,4]");
			}
		}

		WHEN("calling dequeue() until empty") {
			THEN("elements are returned from the front in order") {
				REQUIRE(list.dequeue() == 1);
				REQUIRE(list.dequeue() == 2);
				REQUIRE(list.dequeue() == 3);
				REQUIRE(list.dequeue() == 4);

			}
		}

	}

	GIVEN("a Queue object with one element") {
		Queue list;
		list.enqueue(1);

		WHEN("calling dequeue") {
			int result = list.dequeue();

			THEN("the first element is removed and returned") {
				REQUIRE(result == 1);
				REQUIRE("[]");
			}
		}
	}

	GIVEN("a Queue object with no elements") {
		Queue list;

		WHEN("calling dequeue") {
			int result = list.dequeue();

			THEN("-1 is returned and no crash") {
				REQUIRE(result == -1);
				REQUIRE("[]");
			}
		}
	}
}

SCENARIO("deleting a Queue object on the heap") {
	GIVEN("a Queue object on the heap with some elements") {
		Queue *list = new Queue();

		list->enqueue(1);
		list->enqueue(2);
		list->enqueue(3);
		list->enqueue(4);

		WHEN("deleting") {
			delete list;
			list = NULL;

			THEN("No crash") {
				REQUIRE(1 == 1);
			}
		}
	}
}

SCENARIO("testing Queue assignment operator") {
	GIVEN("a Queue Object with some elements and an empty Queue Object") {
		Queue listA;
		Queue listB;

		listA.enqueue(1);
		listA.enqueue(2);
		listA.enqueue(3);
		listA.enqueue(4); // 1 2 3 4

		WHEN("assigning list with some elements to empty list") {
			listB = listA;

			THEN("both lists contain the same elements and are not empty") {
				REQUIRE(getOutput(listA) == "[1,2,3,4]");
				REQUIRE(getOutput(listB) == "[1,2,3,4]");
			}

			AND_WHEN("one list is modified") {
				listA.dequeue();

				THEN("the other is not") {
					REQUIRE(getOutput(listA) == "[2,3,4]");
					REQUIRE(getOutput(listB) == "[1,2,3,4]");
				}
			}
		}

		WHEN("chain linking assignment operator") {
			Queue listC;

			listC = listB = listA;

			THEN("all lists have the same data") {
				REQUIRE(getOutput(listA) == "[1,2,3,4]");
				REQUIRE(getOutput(listB) == "[1,2,3,4]");
				REQUIRE(getOutput(listC) == "[1,2,3,4]");
			}
		}

		WHEN("assigning an empty list to a non empty list") {
			listA = listB;

			THEN("both lists become empty") {
				REQUIRE(getOutput(listA) == "[]");
				REQUIRE(getOutput(listB) == "[]");
			}
		}

		WHEN("assigning an empty list to an empty list") {
			Queue listC;

			listC = listB;

			THEN("nothing happens and no crash") {
				REQUIRE(getOutput(listB) == "[]");
				REQUIRE(getOutput(listC) == "[]");
			}
		}

		WHEN("assigning a list to itself") {
			Queue listC;

			listA = listA;

			THEN("nothing happens and no crash") {
				REQUIRE(getOutput(listA) == "[1,2,3,4]");
			}
		}
	}
}

SCENARIO("testing Queue::isEmpty()") {

	GIVEN("a Queue object with one element") {
		Queue list;
		list.enqueue(1);

		WHEN("calling isEmpty()") {
			bool result = list.isEmpty();

			THEN("return false") {
				REQUIRE(result == false);
			}
		}
	}

	GIVEN("a Queue object with some elements") {
		Queue list;
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4); // 1 2 3 4

		WHEN("calling isEmpty()") {
			bool result = list.isEmpty();

			THEN("return false") {
				REQUIRE(result == false);
			}
		}
	}

	GIVEN("an empty Queue object") {
		Queue list;
		WHEN("calling isEmpty()") {
			bool result = list.isEmpty();

			THEN("return true") {
				REQUIRE(result == true);
			}
		}
	}

}

SCENARIO("testing Queue::front()") {

	GIVEN("A Queue with one element") {
		Queue list;
		list.enqueue(1);

		WHEN("calling front()") {
			int result = list.front();

			THEN("the front element is returned and list not modified") {
				REQUIRE(result == 1);
				REQUIRE(getOutput(list) == "[1]");
			}
		}
	}

	GIVEN("a Queue object with some elements") {
		Queue list;
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4); // 1 2 3 4

		WHEN("calling front") {
			int result = list.front();

			THEN("the first element is returned and list not modified") {
				REQUIRE(result == 1);
				REQUIRE(getOutput(list) == "[1,2,3,4]");
			}
		}
	}
}

SCENARIO("testing Queue::~Queue") {
	GIVEN("a Queue with no elements") {
		Queue *list = new Queue();

		WHEN("deleting list") {
			delete list;

			THEN("no crash") {
				REQUIRE(1 == 1);
			}
		}
	}

	GIVEN("a Queue with some elements") {
		Queue *list = new Queue();

		list->enqueue(1);
		list->enqueue(2);
		list->enqueue(3);
		list->enqueue(4);

		WHEN("deleting list") {
			delete list;

			THEN("no crash") {
				REQUIRE(1 == 1);
			}
		}
	}

	GIVEN("a Queue with one element") {
		Queue *list = new Queue();

		list->enqueue(1);

		WHEN("deleting list") {
			delete list;

			THEN("no crash") {
				REQUIRE(1 == 1);
			}
		}
	}
}

TEST_CASE("testing complex operations on Queue object") {
	Queue list;

	list.enqueue(-5);
	REQUIRE(getOutput(list) == "[]");
	REQUIRE(list.isEmpty() == true);
	REQUIRE(list.front() == -1);

	list.enqueue(1);
	REQUIRE(getOutput(list) == "[1]");
	REQUIRE(list.isEmpty() == false);
	REQUIRE(list.front() == 1);

	list.enqueue(-69);
	REQUIRE(getOutput(list) == "[1]");

	list.enqueue(2);
	REQUIRE(getOutput(list) == "[1,2]");

	list.enqueue(3);
	REQUIRE(getOutput(list) == "[1,2,3]");

	list.enqueue(4);
	REQUIRE(getOutput(list) == "[1,2,3,4]");

	Queue copyList;
	copyList = list;

	REQUIRE(list.dequeue() == 1);
	REQUIRE(getOutput(list) == "[2,3,4]");

	REQUIRE(list.dequeue() == 2);
	REQUIRE(getOutput(list) == "[3,4]");
	REQUIRE(list.front() == 3);
	REQUIRE(list.dequeue() == 3);
	REQUIRE(getOutput(list) == "[4]");

	REQUIRE(list.dequeue() == 4);
	REQUIRE(getOutput(list) == "[]");
	REQUIRE(list.isEmpty() == true);
	REQUIRE(list.front() == -1);

	REQUIRE(getOutput(copyList) == "[1,2,3,4]");
}

File: archive/cos212/cos212_assignment1/tests_Queue.java
/*
MIT License

Copyright (c) 2017 Evert Geldenhuys

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

public class tests_Queue extends UnitTest{

	public boolean run() {

		boolean result = true;
		result = result & test_enqueu_toString("testing enqueue and toString");
		result = result & test_dequeue("testing dequeue()");
		result = result & test_clone("testing clone");
		result = result & test_complex("test_complex");
		result = result & test_front("test_front");
		result = result & test_isEmpty("test_isEmpty");

		return result;
	}

	public boolean test_template(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END

		// TEST CASE HEADER
		printSubTest("");
		subResult = true;

		// TEST AND ASSERT
		Queue list = new Queue();
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4);

		subResult = subResult & assertCustom(list.toString(), "[4,3,2,1]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END
	}

	public boolean test_enqueu_toString(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END

		// TEST CASE HEADER
		printSubTest("adding positive elements to empty list");
		subResult = true;

		// TEST AND ASSERT
		Queue list = new Queue();
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4); // 1 2 3 4

		subResult = assertCustom(list.toString(), "[1,2,3,4]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// TEST CASE HEADER
		printSubTest("adding negative elements to list");
		subResult = true;

		// TEST AND ASSERT
		list = new Queue();
		list.enqueue(-5);
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(-5);
		list.enqueue(3);
		list.enqueue(4);
		list.enqueue(-5);

		subResult = assertCustom(list.toString(), "[1,2,3,4]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END
	}

	public boolean test_dequeue(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END

		// TEST CASE HEADER
		printSubTest("dequeue elements when list is empty");
		subResult = true;

		// TEST AND ASSERT
		Queue list = new Queue();
		int data = list.dequeue();

		subResult = assertCustom(list.toString(), "[]");
		subResult = subResult & assertCustom(data, -1);

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// TEST CASE HEADER
		printSubTest("deleting elements until list is empty");
		subResult = true;

		// TEST AND ASSERT
		list = new Queue();

		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3); // 1 2 3

		subResult = subResult & assertCustom(list.toString(), "[1,2,3]");

		subResult = subResult & assertCustom(list.dequeue(), 1);
		subResult = subResult & assertCustom(list.toString(), "[2,3]");


		subResult = subResult & assertCustom(list.dequeue(), 2);
		subResult = subResult & assertCustom(list.toString(), "[3]");


		subResult = subResult & assertCustom(list.dequeue(), 3);
		subResult = subResult & assertCustom(list.toString(), "[]");


		subResult = subResult & assertCustom(list.dequeue(), -1);
		subResult = subResult & assertCustom(list.toString(), "[]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// TEST CASES END

		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END
	}

	public boolean test_clone(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END

		// TEST CASE HEADER
		printSubTest("Assigning list to each other");
		subResult = true;

		// TEST AND ASSERT
		Queue list = new Queue();
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4); // 1 2 3 4

		Queue listEmpty = new Queue();

		Queue list2 = new Queue();
		list2.enqueue(5);
		list2.enqueue(6);
		list2.enqueue(7);
		list2.enqueue(8); // 5 6 7 8

		subResult = subResult & assertCustom(list.toString(), "[1,2,3,4]");
		subResult = subResult & assertCustom(list2.toString(), "[5,6,7,8]");
		subResult = subResult & assertCustom(listEmpty.toString(), "[]");

		listEmpty = list.clone();

		subResult = subResult & assertCustom(list.toString(), "[1,2,3,4]");
		subResult = subResult & assertCustom(listEmpty.toString(), "[1,2,3,4]");

		list.dequeue();
		list.enqueue(55);
		list.enqueue(44);

		subResult = subResult & assertCustom(list.toString(), "[2,3,4,55,44]");
		subResult = subResult & assertCustom(listEmpty.toString(), "[1,2,3,4]");

		listEmpty = list2.clone();

		subResult = subResult & assertCustom(list2.toString(), "[5,6,7,8]");
		subResult = subResult & assertCustom(listEmpty.toString(), "[5,6,7,8]");

		listEmpty = listEmpty.clone();
		subResult = subResult & assertCustom(listEmpty.toString(), "[5,6,7,8]");

		listEmpty = new Queue();
		list2 = listEmpty.clone();

		subResult = subResult & assertCustom(list2.toString(), "[]");
		subResult = subResult & assertCustom(listEmpty.toString(), "[]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END

	}

	public boolean test_complex(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END

		// TEST CASE HEADER
		printSubTest("Mixing operations");
		subResult = true;

		// TEST AND ASSERT
		Queue list = new Queue();
		list.enqueue(-1);
		list.enqueue(-5);
		list.enqueue(2);
		list.enqueue(-31);
		list.enqueue(3);
		list.enqueue(-133);
		list.enqueue(-13);
		list.enqueue(4);

		subResult = subResult & assertCustom(list.toString(), "[2,3,4]");

		subResult = subResult & assertCustom(list.dequeue(), 2);
		subResult = subResult & assertCustom(list.dequeue(), 3);
		subResult = subResult & assertCustom(list.dequeue(), 4);
		subResult = subResult & assertCustom(list.dequeue(), -1);

		list.enqueue(5);
		subResult = subResult & assertCustom(list.toString(), "[5]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END
	}

	public boolean test_front(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END


		// TEST CASE HEADER
		printSubTest("front no element");
		subResult = true;

		// TEST AND ASSERT
		Queue list = new Queue();

		subResult = subResult & assertCustom(list.front(), -1);

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);


		// TEST CASE HEADER
		printSubTest("front one element");
		subResult = true;

		// TEST AND ASSERT
		list = new Queue();
		list.enqueue(5);

		subResult = subResult & assertCustom(list.front(), 5);

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);


		// TEST CASE HEADER
		printSubTest("front many elements");
		subResult = true;

		// TEST AND ASSERT
		list = new Queue();
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4); // 1 2 3 4

		subResult = subResult & assertCustom(list.toString(), "[1,2,3,4]");
		subResult = subResult & assertCustom(list.front(), 1);
		subResult = subResult & assertCustom(list.toString(), "[1,2,3,4]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END
	}

	public boolean test_isEmpty(String title) {
		// HEADER
		printHeader(title);
		boolean result = true;
		boolean subResult = true;
		// HEADER END


		// TEST CASE HEADER
		printSubTest("no element");
		subResult = true;

		// TEST AND ASSERT
		Queue list = new Queue();

		subResult = subResult & assertCustom(list.isEmpty(), true);

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);


		// TEST CASE HEADER
		printSubTest("one element");
		subResult = true;

		// TEST AND ASSERT
		list = new Queue();
		list.enqueue(5);

		subResult = subResult & assertCustom(list.isEmpty(), false);

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);


		// TEST CASE HEADER
		printSubTest("many elements");
		subResult = true;

		// TEST AND ASSERT
		list = new Queue();
		list.enqueue(1);
		list.enqueue(2);
		list.enqueue(3);
		list.enqueue(4); // 1 2 3 4

		subResult = subResult & assertCustom(list.toString(), "[1,2,3,4]");
		subResult = subResult & assertCustom(list.isEmpty(), false);
		subResult = subResult & assertCustom(list.toString(), "[1,2,3,4]");

		// TEST CASE FOOTER
		result = result & subResult;
		printSubTestFooter(subResult);

		// FOOTER
		printFooter(title, result);
		return result;
		// FOOTER END
	}
}
